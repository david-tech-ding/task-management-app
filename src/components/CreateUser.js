import React, { useState, useRef } from "react";
import { useNavigate } from "react-router-dom";
import axios from "axios";
import "../styles/create-user.css";

const CreateUser = ({ onSetLoggedInUser, user }) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [jobRole, setJobRole] = useState("");
  const navigate = useNavigate();

  const userRef = useRef();
  const formRef = useRef();

  const handleEnterKeyPress = (e) => {
    if (e.keyCode === 13) {
      e.preventDefault();
      handleCreateUser(e);
    }
  };

  const addEnterKeyPressListener = () => {
    if (userRef.current) {
      userRef.current.addEventListener("keydown", handleEnterKeyPress);
    }
    if (formRef.current) {
      formRef.current.addEventListener("keydown", handleEnterKeyPress);
    }
  };
  const removeEnterKeyPressListener = () => {
    if (userRef.current) {
      userRef.current.removeEventListener("keydown", handleEnterKeyPress);
    }
    if (formRef.current) {
      formRef.current.removeEventListener("keydown", handleEnterKeyPress);
    }
  };

  const handleFirstNameChange = (e) => {
    setFirstName(e.target.value);
  };

  const handleLastNameChange = (e) => {
    setLastName(e.target.value);
  };

  const handleJobRoleChange = (e) => {
    setJobRole(e.target.value);
  };

  const handleCreateUser = (e) => {
    e.preventDefault();
    removeEnterKeyPressListener();
    if (firstName && lastName && jobRole) {
      axios
        .patch(`/user/id/${user.id}`, {
          firstName,
          lastName,
          jobRole,
        })
        .then(() => {
          onSetLoggedInUser({ ...user, firstName: firstName });
          setFirstName("");
          setLastName("");
          setJobRole("");
        })
        .catch((err) => {
          console.log("Error creating user:", err);
        });
      navigate("/");
    } else {
      alert("Please fill in the required fields");
    }
    addEnterKeyPressListener();
  };
  return (
    <div className="create-user">
      <h2>Finish setting up your account:</h2>
      <form onSubmit={handleCreateUser} ref={formRef}>
        <div>
          <label htmlFor="firstName">First Name:</label>
          <input
            type="text"
            id="firstName"
            value={firstName}
            onChange={handleFirstNameChange}
          />
        </div>
        <div>
          <label htmlFor="lastName">Last Name:</label>
          <input
            type="text"
            id="lastName"
            value={lastName}
            onChange={handleLastNameChange}
          />
        </div>
        <div>
          <label htmlFor="jobRole">Job Role:</label>
          <input
            type="text"
            id="jobRole"
            value={jobRole}
            onChange={handleJobRoleChange}
          />
        </div>
        <button onClick={handleCreateUser}>Create User</button>
      </form>
    </div>
  );
};

export default CreateUser;
