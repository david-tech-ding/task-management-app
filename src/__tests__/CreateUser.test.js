import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import CreateUser from "../components/CreateUser";

const onSetLoggedInUser = jest.fn();

describe("renders", () => {
  const validProps = {
    id: 1,
    userName: "test",
    firstName: "test",
  };
  it("renders correctly", () => {
    const { asFragment } = render(
      <Router>
        <CreateUser onSetLoggedInUser={onSetLoggedInUser} user={validProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
