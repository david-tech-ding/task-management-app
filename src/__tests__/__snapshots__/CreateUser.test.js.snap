// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders renders correctly 1`] = `
<DocumentFragment>
  <div
    class="create-user"
  >
    <h2>
      Finish setting up your account:
    </h2>
    <form>
      <div>
        <label
          for="firstName"
        >
          First Name:
        </label>
        <input
          id="firstName"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="lastName"
        >
          Last Name:
        </label>
        <input
          id="lastName"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="jobRole"
        >
          Job Role:
        </label>
        <input
          id="jobRole"
          type="text"
          value=""
        />
      </div>
      <button>
        Create User
      </button>
    </form>
  </div>
</DocumentFragment>
`;
